### PR TITLE
#16: Implement JwtService

### LINKED ISSUE #16: JWT token service

### PR DESCRIPTION
**Pull request by ilisau**
In this PR I implemented JwtService for user authentication and close #16

### CODE REVIEW

**Comment by ilisau** [comment]
Should I edit workflow file, because there are export constants and I have added some new of them in project?

**Comment by h1alexbel** [comment]
@ilisau shure, go ahead

**Comment by h1alexbel** [comment]
@ilisau please fix compilation failure

File: src/main/java/com/taskviewer/api/web/security/jwt/JwtUserDetailsFactory.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+public class JwtUserDetailsFactory {
+
+  public static JwtUserDetails create(User user) {
+    return new JwtUserDetails(
+            user.id(),
+            user.username(),
+            user.email(),
+            null,
+            true,
+            mapToGrantedAuthority(Collections.singletonList(user.role())));
```
Comment (h1alexbel) [inline src/main/java/com/taskviewer/api/web/security/jwt/JwtUserDetailsFactory.java:22]: line is too long

File: src/main/java/com/taskviewer/api/web/security/jwt/JwtUserDetailsFactory.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+import io.jsonwebtoken.Claims;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+public class JwtUserDetailsFactory {
```
Comment (h1alexbel) [inline src/main/java/com/taskviewer/api/web/security/jwt/JwtUserDetailsFactory.java:13]: I recommend a better design:
```java
public class JwtUserDetails {

private final User user;

// methods
}
```
Thus we will encapsulate the user and reuse it

File: src/main/java/com/taskviewer/api/web/security/jwt/JwtUserDetailsFactory.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+            true,
+            mapToGrantedAuthority(Collections.singletonList(user.role())));
+  }
+
+  public static JwtUserDetails create(Claims claims) {
+    String role = claims.get("role", String.class);
+    return new JwtUserDetails(
+            claims.get("id", Long.class),
+            claims.getSubject(),
+            null,
```
Comment (h1alexbel) [inline src/main/java/com/taskviewer/api/web/security/jwt/JwtUserDetailsFactory.java:30]: don't use `NULL`

File: src/main/java/com/taskviewer/api/web/security/jwt/JwtUserDetailsFactory.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+            null,
+            true,
+            mapToGrantedAuthority(Collections.singletonList(role)));
+  }
+
+
+  private static List<GrantedAuthority> mapToGrantedAuthority(List<String> userRoles) {
```
Comment (h1alexbel) [inline src/main/java/com/taskviewer/api/web/security/jwt/JwtUserDetailsFactory.java:37]: each private method can be represented as new class, thus we can reuse it functionality and test it to full extent

File: src/main/java/com/taskviewer/api/web/security/jwt/JwtUserDetailsService.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+@Service
+@RequiredArgsConstructor
+public class JwtUserDetailsService implements UserDetailsService {
+
+  private final Users users;
+
+  @Override
+  public UserDetails loadUserByUsername(String username) {
+    UserNotFoundException exception = new UserNotFoundException("User with username %s is not found".formatted(username));
```
Comment (h1alexbel) [inline src/main/java/com/taskviewer/api/web/security/jwt/JwtUserDetailsService.java:19]: split the line, it's too long

File: src/main/java/com/taskviewer/api/web/security/jwt/JwtUserDetailsFactory.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+import java.util.Collections;
+
+public class JwtUserDetailsFactory {
+
+  public static JwtUserDetails create(User user) {
+    return new JwtUserDetails(user);
+  }
+
+  public static JwtUserDetails create(Claims claims) {
```
Comment (h1alexbel) [inline src/main/java/com/taskviewer/api/web/security/jwt/JwtUserDetailsFactory.java:14]: I mean this class will encapsulates the `user`, not pass it as a param, but encapsulates.
and you will get rid of these static methods

File: src/main/java/com/taskviewer/api/web/security/jwt/JwtUserDetailsFactory.java [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    String role = claims.get("role", String.class);
+    return new JwtUserDetails(
+            claims.get("id", Long.class),
+            claims.getSubject(),
+            GrantedAuthorityMapper.mapToGrantedAuthority(Collections.singletonList(role)));
```
Comment (h1alexbel) [inline src/main/java/com/taskviewer/api/web/security/jwt/JwtUserDetailsFactory.java:19]: this line is too long. split it please

**Review by h1alexbel — CHANGES_REQUESTED** [review]

**Review by h1alexbel — APPROVED** [review]

**Comment by h1alexbel** [comment]
@rultor merge

**Comment by h1alexbel** [comment]
@ilisau thanks!

**Pull request merged**